improvement(client): improved internal submitMessage, reSubmit, rollback, and submitSignal type-safety (#24319)
microsoft/FluidFramework: .changeset/frowny-paths-hear.md

> ---
> "@fluidframework/datastore": minor
> ---
> ---
> "section": legacy
> ---

Reviewer: We're in a transition period for changesets, this is the new format, with Fluid-specific settings in the same front matter "container", prefixed with double underscore.
```suggestion
---
"@fluidframework/datastore": minor
"__section": legacy
---
```

> Deprecate `FluidDataStoreRuntime.submitMessage`

Reviewer: We avoid formatting in headings.
```suggestion
Deprecate FluidDataStoreRuntime.submitMessage
```

> `FluidDataStoreRuntime.submitMessage` implementation is not required per `IFluidDataStoreChannel` and is now deprecated. `IFluidParentContext` (which is base interface for `IFluidDataStoreContext`) should always be used to access `submitMessage` functionality.

Reviewer: ```suggestion
Implementing `FluidDataStoreRuntime.submitMessage` is not required per `IFluidDataStoreChannel` and is now deprecated. `IFluidParentContext` (the base interface for `IFluidDataStoreContext`) should always be used to access `submitMessage` functionality.
```

---
"@fluidframework/datastore": minor
"@fluidframework/test-runtime-utils": minor
"__section": legacy
---

Deprecate submitMessage on FluidDataStoreRuntime and MockFluidDataStoreRuntime

Implementing `FluidDataStoreRuntime.submitMessage` is not required per `IFluidDataStoreChannel` and is now deprecated on `FluidDataStoreRuntime` and corresponding `MockFluidDataStoreRuntime`.

> See [issue #24406](https://github.com/microsoft/FluidFramework/issues/24406) for details and alternatives.

Reviewer 2: That issue mentions:

> IFluidParentContext (the base interface for IFluidDataStoreContext) should always be used to access submitMessage functionality.

That may be true for _inside_ `FluidDataStoreRuntime` class, but as for the public API to submit an op, I think the comment should just be - "Submitting arbitrary messages is not supported"

Author: Is `FluidDataStoreRuntime` a class customers are expected to derive from? I don't think Pages does and I see that we do internally in at least two places.
Feel free edit the issue.
We should seal the class if no one should derive from it. Then it will be easier to blanketly say don't call.
